### PR TITLE
Add Virtual scrolling / rendering to the Menu

### LINKED ISSUE #983: Update Menu to use virtual render window to allow rendering of large option sets
**Enhancement** 

We should enhance the menu widget to use a render window similar to that used by grid. This would allow it to accept many thousands of options and be able to render them quickly and effectively.

### PR DESCRIPTION
**Pull request by tomdye**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/framework/blob/master/STYLE.md)
* [x] Unit tests are included in the PR
* [x] For new widgets, an entry has been added to the `.dojorc`
* [x] Any widget variant uses `theme.compose` like [this](https://github.com/dojo/widgets/issues/847)
* [x] WidgetProperties are exported

**Description:**

- Adds scrolling virtual window
- adds required `total` property required
   - This will be used for future enhancements around pagination

Resolves #983

### CODE REVIEW

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-5c9o1diht.now.sh
* Tests Deployment: https://dojowidgets-of5t0kgb1.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-7kr8opgls.now.sh
* Tests Deployment: https://dojowidgets-hkf7khg8g.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-4wtwsxcl2.now.sh
* Tests Deployment: https://dojowidgets-5h3nvwy5z.now.sh

File: src/examples/src/config.tsx [hunk 1]
```diff
@@ -126,6 +126,7 @@ import FocusTooltip from './widgets/tooltip/Focus';
 
 `!has('docs')`;
 import testsContext from './tests';
+import LargeOptionSet from './widgets/menu/LargeOptionSet';
```
Comment (agubler) [inline src/examples/src/config.tsx]: Is this supposed to be at the bottom?

File: src/examples/src/config.tsx [hunk 2]
```diff
@@ -502,6 +503,12 @@ export const config = {
 					filename: 'ListBox',
 					module: ListBox,
 					title: 'List Box'
+				},
+				{
+					description: 'This example shows the menu rendering 100,000 options.',
```
Comment (agubler) [inline src/examples/src/config.tsx]: It's not really rendering 100,000 options, perhaps we can better describe it.

File: src/examples/src/widgets/menu/LargeOptionSet.tsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { create, tsx } from '@dojo/framework/core/vdom';
+import Menu, { MenuOption } from '@dojo/widgets/menu';
+import icache from '@dojo/framework/core/middleware/icache';
+
+const factory = create({ icache });
+
+export default factory(function LargeOptionSet({ middleware: { icache } }) {
+	const options: MenuOption[] = [];
+	const total = 100000;
+	for (let i = 0; i < total; i++) {
```
Comment (agubler) [inline src/examples/src/widgets/menu/LargeOptionSet.tsx]: I don't think you need to do this every render?

File: src/menu/MenuItem.tsx [hunk 4]
```diff
@@ -13,43 +10,16 @@ export interface MenuItemProperties {
...
-		disabled = false,
-		widgetId
-	} = properties();
+const factory = create({ theme }).properties<MenuItemProperties>();
 
-	if (icache.get('active') !== active) {
-		icache.set('active', active);
-		active && onActive(dimensions.get('root'));
-	}
+export const MenuItem = factory(function({ properties, children, middleware: { theme } }) {
```
Comment (agubler) [inline src/menu/MenuItem.tsx]: `function MenuItem`

File: src/menu/index.tsx [hunk 5]
```diff
@@ -79,70 +81,27 @@ const offscreenHeight = (dnode: RenderResult) => {
...
-	dimensions,
 	theme
 }).properties<MenuProperties>();
 
-export const Menu = factory(function({
-	properties,
-	id,
-	middleware: { icache, focus, dimensions, theme }
-}) {
+export const Menu = factory(function({ properties, id, middleware: { icache, focus, theme } }) {
```
Comment (agubler) [inline src/menu/index.tsx]: `function Menu`

File: src/menu/index.tsx [hunk 6]
```diff
@@ -59,10 +60,11 @@ interface MenuICache {
 	inputText: string;
 	itemHeight: number;
 	itemsInView: number;
-	itemToScroll: number;
 	menuHeight: number;
 	resetInputTextTimer: any;
 	value: string;
+	scrollTop: number;
```
Comment (agubler) [inline src/menu/index.tsx]: Don't these properties require comments?

File: src/menu/index.tsx [hunk 7]
```diff
@@ -213,23 +183,137 @@ export const Menu = factory(function({
...
+		const menuItemChild = itemRenderer
+			? itemRenderer({
+					selected: false,
+					active: false,
+					value: 'offscreen',
+					disabled: false
+			  })
+			: 'offscreen';
+
+		const offscreenMenuItem = <MenuItem {...offscreenItemProps}>{menuItemChild}</MenuItem>;
```
Comment (agubler) [inline src/menu/index.tsx]: What if it's a `ListBoxItem`?

File: src/examples/src/config.tsx (hunk 1 above)
Comment (tomdye) [reply src/examples/src/config.tsx]: good catch

File: src/examples/src/config.tsx (hunk 2 above)
Comment (tomdye) [reply src/examples/src/config.tsx]: sure

File: src/examples/src/widgets/menu/LargeOptionSet.tsx (hunk 3 above)
Comment (tomdye) [reply src/examples/src/widgets/menu/LargeOptionSet.tsx]: fair

File: src/menu/index.tsx (hunk 7 above)
Comment (tomdye) [reply src/menu/index.tsx]: ATM they're the same size etc, only difference is the aria / role tags and the ability to select them

Comment (agubler) [reply src/menu/index.tsx]: Aren't they styled by the theme? So in theory could be different sizes if someone themes them like that?

Comment (tomdye) [reply src/menu/index.tsx]: fair point, good catch, thanks

File: src/menu/index.tsx (hunk 6 above)
Comment (tomdye) [reply src/menu/index.tsx]: no, these are internal icache interfaces

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-c3hbs2mp3.now.sh
* Tests Deployment: https://dojowidgets-fcnm9ay0q.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-6a76mzqqx.now.sh
* Tests Deployment: https://dojowidgets-69z1x7ufr.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-3gqfeg07u.now.sh
* Tests Deployment: https://dojowidgets-ev1ranaop.now.sh

**Pull request merged**